awx_collection/plugins/modules/tower_team.py

```diff
@@ -57,57 +67,60 @@
     )
 
+    # Create a module for ourselves
     module = TowerModule(argument_spec=argument_spec, supports_check_mode=True)
 
+    # Extract our parameters
     name = module.params.get('name')
+    new_name = module.params.get('new_name')
     description = module.params.get('description')
     organization = module.params.get('organization')
     state = module.params.get('state')
 
-    json_output = {'team': name, 'state': state}
-
-    tower_auth = tower_auth_config(module)
-    with settings.runtime_values(**tower_auth):
-        tower_check_mode(module)
-        team = tower_cli.get_resource('team')
-
-        try:
-            org_res = tower_cli.get_resource('organization')
-            org = org_res.get(name=organization)
-
-            if state == 'present':
-                result = team.modify(name=name, organization=org['id'],
-                                     description=description, create_on_missing=True)
-                json_output['id'] = result['id']
-            elif state == 'absent':
-                result = team.delete(name=name, organization=org['id'])
-        except (exc.NotFound) as excinfo:
-            module.fail_json(msg='Failed to update team, organization not found: {0}'.format(excinfo), changed=False)
-        except (exc.ConnectionError, exc.BadRequest, exc.AuthError) as excinfo:
-            module.fail_json(msg='Failed to update team: {0}'.format(excinfo), changed=False)
-
-    json_output['changed'] = result['changed']
-    module.exit_json(**json_output)
+    # Attempt to lookup the related items the user specified (these will fail the module if not found)
+    org_id = module.resolve_name_to_id('organizations', organization)
+
+    # Attempt to lookup team based on the provided name and org ID
+    team = module.get_one('teams', **{
+        'data': {
+            'name': name,
+            'organization': org_id
+        }
+    })
+
+    # Create data to sent to create and update
+    team_fields = {
+        'name': new_name if new_name else name,
+        'description': description,
+        'organization': org_id
+    }
+
+    if state == 'absent' and not team:
+        # If the state was absent and we had no team, we can just return
+        module.exit_json(**module.json_output)
+    elif state == 'absent' and team:
+        # If the state was absent and we had a team, we can try to delete it, the module will handle exiting from this
+        module.delete_endpoint('teams/{0}'.format(team['id']), item_type='team', item_name=name, **{})
+    elif state == 'present' and not team:
+        # if the state was present and we couldn't find a team we can build one, the module will handle exiting from this
+        module.post_endpoint('teams', item_type='team', item_name=name, **{'data': team_fields})
+    else:
+        # If the state was present and we had a team we can see if we need to update it
+        # This will return on its own
+        module.update_if_needed(team, team_fields)
```

Comment:
As a specific example, following up to my last comment, both `delete_endpoint` and `post_endpoint` need the plural name used in the list endpoint, the string `"teams"` here. The method `update_if_needed` avoids this by using the URL in the data, but it doesn't need to. It could also do a PATCH to `'teams/{0}'.format(team['id'])`, and I would prefer this because it's more consistent.

The arg `item_name` doesn't seem to me like something we need to keep. We could default to looking for name, and then have some other means of overriding that on initialization of the class for the special cases.

The data is clearly needed in all cases.

So this points us in the direction of a call pattern like `'teams', team_fields`, and you could share that pattern between all the methods from module_utils.